README.md
LineIRCd
========

fork from UnrealIRCd for ircline.ru network


Changes
/*
 * UnrealIRCd Changes File - (C) Carsten Munk 1999-2002 &
 *                               The UnrealIRCd Team
 *                              
 * $Id: Changes,v 1.1.1.1.2.17 2004/07/03 19:04:02 syzop Exp $
 *                              
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 1, or (at your option)
 * any later version.           
 *                              
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *                              
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *                              
 */                             
                                
[Unreal 3.2.1]
- Replaced tre.dll/tre.lib, previous versions caused a crash (eg: if you included
  spamfilter.conf).
- Fixed a problem when compiling with GUEST enabled. Reported by dvzion (#0001758)
- Fixed a documentation typo reported by Konc
- Fixed a little problem in ./unreal to move ircd.pid.bak to ircd.pid of starting failed.
  Fixed by fez (#0001739)
- Made it so chg* command usage is not logged from U:lines. Reported by diskman1 (#0001718)
- Changed int_to_base64() warning so it has less false positives (#0001797).
- Fixed a bug where if set::ssl::options was empty, a crash could occur. Reported by
  spider312 (#0001799)
- Include openssl/md5.h and openssl/ripemd.h if compiled w/SSL, this seems how it
  should be done and also makes unreal w/SSL able to compile on OpenBSD (3.5).
- Added module support for Windows!
  To module coders: 
  -Building your module on Windows is almost exactly the same as on *nix
  nmake -f makefile.win32 custommodule MODULEFILE=thefile

  -Also, as I'm sure most module developers will begin distributing .dll files for their
  modules, it is important to note that just like the modules on *nix, the dlls will need
  to be recompiled each time a new version of Unreal is released. This is simply due to
  the potential for binary incompatibility which at this point can not be avoided.

  -Most modules already do this, but it is now a requirement for Windows support. At the
  very minimum, the Mod_Test function must have the DLLFUNC specifier in order to work
  Unreal will automatically add DLLFUNC to the other functions, but it does not hurt if
  you use it for all of them.

  NOTE: It is expected that there will probably be a few bugs to work out with this, so
  please consider testing it! And module coders, please report any problems you experience!
- Modulized cloaking:
  - Currently there are 2 cloaking modules: 'cloak' and 'oldcloak'.
  - You MUST load at least 1 of them (like: loadmodule "src/modules/cloak.so";).
  - 'oldcloak' is just the <=3.2 cloaking algorithm.
  - 'cloak' is the new, recommended, and much more secure cloaking algorithm that uses
    md5 internally and requires 3 keys of 5-100 characters (10-20 is fine) consisting of
    mixed lowcase (a-z), upcase (A-Z) and digits (0-9) [eg: "AopAS6WQH2Os6hfosh4SFJHs"].
  - Note that 'oldcloak' is only ment for use during the upgrade process (it will give
    a warning if you use it), you should switch to the much more secure 'cloak' module
    as soon as all your servers are upgraded.
  - Module coders can code 3rd party cloak modules that use other algorithms or cloak
    in a different way (have a look at src/modules/cloak.c).
- spamfilter: Added Bloodhound.Exploit.6 sig
- Compile fixes for win32 modules with ssl/zip/curl
- Changed 'Services operator' in /whois (back) to 'Services administrator', this was
  requested by many people and seems to be the best after all (#0001634).
- Local opers can now also join +O (operonly) channels (#0001694).
- Changed the way MSG/NOTICE <prefix>#chan works:
  - It now goes to <prefix> and higher, so '/notice +#chan hi!' goes to +vhoaq
  - You need at least voice in order to be able to msg/notice +#chan, %#chan or @#chan
  - You need at least ops in order to be able to msg/notice &#chan or ~#chan
  - Any multi-prefix targets will be converted automatically (eg: ~&@#chan to @#chan).
  - internal: use of the CHANOPPFX macro is now deprecated.
  All of this was done to make it a bit more 'safe' and userfriendly (#0001812).
- Fixed a remote include bug that would cause unnecessary blocking
- Fixed a /credits typo
- Imported TRE 0.6.7 for win32
- Imported TRE 0.6.7 for *nix and made use of tre_version to report the version of TRE
  in use at startup
- Fixed a Win32 module bug
- Added set::spamfilter::virus-help-channel-deny. This allows you to block any
  normal joins to the virus-help-channel. This way you could prevent users into
  accidental (or tricked) joining of the virus-help-channel and becomming infected.
  This feature is disabled by default. Requested by bleepy (#0001811).
- German doc updates (week 20).
- Added optional parameter to SVSJOIN to deal with channel keys. Reported by 
  DukePyrolator (#0001822).
- Added zlib+SSL version check on boot to make sure the runtime version is the same
  as the 'compiled for' (header) version. If they mismatch, UnrealIRCd could crash,
  so a big warning is posted if it happends.
- Improved the above: made it work on windows and also added a check for curl.
- spamfilter.conf: Added yet another sig for a site that causes Backdoor.Delf.lq
  infection (reported by nexus), also changed LOI trojan and Bloodhound.Exploit.6
  action to gline.
- Fixed a win32 module bug that broke strcasecmp/stricmp
- German doc updates (week 21)
- Fixed a permanent modules bug: custom allow/except/ban/deny types were lost

src/zip.c
/*
 *   IRC - Internet Relay Chat, ircd/s_zip.c
 *   Copyright (C) 1996  Christophe Kalt
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 1, or (at your option)
 *   any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   $Id: zip.c,v 1.1.6.12 2004/07/03 19:04:32 syzop Exp $
 */
#include "struct.h"
#include "common.h"
#include "sys.h"
#include "numeric.h"
#include "msg.h"
#include "channel.h"
#include "version.h"
#include "proto.h"

#include <string.h>
#include <stdlib.h>

#ifdef  ZIP_LINKS
/*
** Important note:
**      The provided buffers for uncompression and compression *MUST* be big
**      enough for any operation to complete.
**
**      s_bsd.c current settings are that the biggest packet size is 16k
**      (but socket buffers are set to 8k..)
*/

/*
** size of the buffer holding compressed data
**
** outgoing data:
**      must be enough to hold compressed data resulting of the compression
**      of up to ZIP_MAXIMUM bytes
** incoming data:
**      must be enough to hold cptr->zip->inbuf + what was just read
**      (cptr->zip->inbuf should never hold more than ONE compression block.
**      The biggest block allowed for compression is ZIP_MAXIMUM bytes)
*/
#define ZIP_BUFFER_SIZE         (ZIP_MAXIMUM + READBUF_SIZE)

/*
** size of the buffer where zlib puts compressed data
**      must be enough to hold uncompressed data resulting of the
**      uncompression of zibuffer
**
**      I'm assuming that at best, ratio will be 25%. (tests show that
**      best ratio is around 40%).
*/

/*
 * On an hybrid test net, we kept filling up unzipbuf
 * original was 4*ZIP_BUFFER_SIZE
 *
 * -Dianora
 */

#define UNZIP_BUFFER_SIZE       6 * ZIP_BUFFER_SIZE

/* buffers */
static  char    unzipbuf[UNZIP_BUFFER_SIZE];
static  char    zipbuf[ZIP_BUFFER_SIZE];

/*
** zip_init
**      Initialize compression structures for a server.
**      If failed, zip_free() has to be called.
*/
int     zip_init(aClient *cptr, int compressionlevel)
{
  cptr->zip  = (aZdata *) MyMalloc(sizeof(aZdata));
  cptr->zip->incount = 0;
  cptr->zip->outcount = 0;

  cptr->zip->in  = (z_stream *) MyMalloc(sizeof(z_stream));
  bzero(cptr->zip->in, sizeof(z_stream)); /* Just to be sure -- Syzop */
  cptr->zip->in->total_in = 0;
  cptr->zip->in->total_out = 0;
  cptr->zip->in->zalloc = NULL;
  cptr->zip->in->zfree = NULL;
  cptr->zip->in->data_type = Z_ASCII;
  if (inflateInit(cptr->zip->in) != Z_OK)
    {
      cptr->zip->out = NULL;
      return -1;
    }

  cptr->zip->out = (z_stream *) MyMalloc(sizeof(z_stream));
  bzero(cptr->zip->out, sizeof(z_stream)); /* Just to be sure -- Syzop */
  cptr->zip->out->total_in = 0;
  cptr->zip->out->total_out = 0;
  cptr->zip->out->zalloc = NULL;
  cptr->zip->out->zfree = NULL;
  cptr->zip->out->data_type = Z_ASCII;
  if (deflateInit(cptr->zip->out, compressionlevel) != Z_OK)
    return -1;

  return 0;
}

/*
** zip_free
*/
void    zip_free(aClient *cptr)
{
	ClearZipped(cptr);
	if (cptr->zip)
	{
		if (cptr->zip->in)
			inflateEnd(cptr->zip->in);
		MyFree(cptr->zip->in);
		cptr->zip->in = NULL;
		if (cptr->zip->out) 
			deflateEnd(cptr->zip->out);
		MyFree(cptr->zip->out);
		cptr->zip->out = NULL;
		MyFree(cptr->zip);
		cptr->zip = NULL;
	}
}

/*
** unzip_packet
**      Unzip the buffer,
**      put anything left in cptr->zip->inbuf, update cptr->zip->incount
**
**      will return the uncompressed buffer, length will be updated.
**      if a fatal error occurs, length will be set to -1
*/
char *unzip_packet(aClient *cptr, char *buffer, int *length)
{
  z_stream *zin = cptr->zip->in;
  int   r;
  char  *p;

  if(cptr->zip->incount)
    {
      /* There was a "chunk" of uncompressed data without a newline
       * left over from last unzip_packet. So pick that up, and unzip
       * some more data. Note, buffer parameter isn't used in this case.
       * -Dianora
       */
      memcpy((void *)unzipbuf,(void *)cptr->zip->inbuf,cptr->zip->incount);
      zin->avail_out = UNZIP_BUFFER_SIZE - cptr->zip->incount;
      zin->next_out = (Bytef *) (unzipbuf + cptr->zip->incount);
      cptr->zip->incount = 0;
      cptr->zip->inbuf[0] = '\0'; /* again unnecessary but nice for debugger */
    }
  else
    {
      /* Start unzipping this buffer, if I fill up output buffer,
       * then snag whatever uncompressed incomplete chunk I have at
       * the top of the uncompressed buffer, save it for next pass.
       * -Dianora
       */
      if(!buffer)       /* Sanity test never hurts */
        {
          *length = -1;
          return((char *)NULL);
        }
      zin->next_in = (Bytef *) buffer;
      zin->avail_in = *length;
      zin->next_out = (Bytef *) unzipbuf;
      zin->avail_out = UNZIP_BUFFER_SIZE;
    }

  switch (r = inflate(zin, Z_NO_FLUSH))
    {
    case Z_OK:
      if (zin->avail_in)
        {
          cptr->zip->incount = 0;

          if(zin->avail_out == 0)
            {
              /* ok, filled up output buffer, complain about it, but go on.
               * I need to find any incomplete "chunk" at the top of
               * the uncompressed output buffer, and save it for next call.
               * N.B. That cptr->zip->inbuf isn't really necessary
               * i.e. re-entrancy is not required since I know
               * there is more uncompressed data to do, and dopacket()
               * will not return until its all parsed. -db
               */
              sendto_realops("Overflowed unzipbuf increase UNZIP_BUFFER_SIZE");
              /*
               * I used to just give up here....
               * length = -1;
               * return((char *)NULL);
               */

              /* Check for pathological case where output
               * just happened to have finished with a newline
               * and there is still input to do
               * Just stuff a newline in for now, it will be discarded
               * anyway, and continue parsing. -db
               */

              if((zin->next_out[0] == '\n') || (zin->next_out[0] == '\r'))
                {
                  cptr->zip->inbuf[0] = '\n';
                  cptr->zip->incount = 1;
                }
              else
                {
                  /* Scan from the top of output, look for a complete
                   * "chunk" N.B. there should be a check for p hitting
                   * the bottom of the unzip buffer. -db
                   */

                  for(p = (char *) zin->next_out;p >= unzipbuf;)
                    {
                      if((*p == '\r') || (*p == '\n'))
                        break;
                      zin->avail_out++;
                      p--;
                      cptr->zip->incount++;
                    }
                  /* A little sanity test never hurts -db */
                  if(p == unzipbuf)
                    {
                      cptr->zip->incount = 0;
                      cptr->zip->inbuf[0] = '\0';       /* only for debugger */
                      *length = -1;
                      return((char *)NULL);
                    }
                  /* Ok, stuff this "chunk" into inbuf
                   * for next call -Dianora 
                   */
                  p++;
                  cptr->zip->incount--;
                  memcpy((void *)cptr->zip->inbuf,
                         (void *)p,cptr->zip->incount);
                }
            }
          else
            {
              /* outbuf buffer is not full, but still
               * input to do. I suppose its just bad data.
               * However I don't have much other choice here but to
               * give up in complete disgust -db
               */
              *length = -1;
              return((char *)NULL);
            }
        }

      *length = UNZIP_BUFFER_SIZE - zin->avail_out;
      return unzipbuf;

    case Z_BUF_ERROR:
      if (zin->avail_out == 0)
        {
          sendto_realops("inflate() returned Z_BUF_ERROR: %s",
                      (zin->msg) ? zin->msg : "?");
          *length = -1;
        }
      break;

    case Z_DATA_ERROR: /* the buffer might not be compressed.. */
      if (!strncmp("ERROR ", buffer, 6))
        {
          cptr->zip->first = 0;
          ClearZipped(cptr);
          /*
           * This is not sane at all.  But if other server
           * has sent an error now, it is probably closing
           * the link as well.
           */
          return buffer;
        }
        /* Let's be nice and give them a hint ;) -- Syzop */
        sendto_realops("inflate() error: * Are you perhaps linking zipped with non-zipped? *");
        sendto_realops("Hint: link::options::zip should be the same at both sides (either both disabled or both enabled)");
        /* no break */

    default: /* error ! */
      /* should probably mark link as dead or something... */
      sendto_realops("inflate() error(%d): %s", r,
                  (zin->msg) ? zin->msg : "?");
      *length = -1; /* report error condition */
      break;
    }
  return((char *)NULL);
}

/*
** zip_buffer
**      Zip the content of cptr->zip->outbuf and of the buffer,
**      put anything left in cptr->zip->outbuf, update cptr->zip->outcount
**
**      if flush is set, then all available data will be compressed,
**      otherwise, compression only occurs if there's enough to compress,
**      or if we are reaching the maximum allowed size during a connect burst.
**
**      will return the uncompressed buffer, length will be updated.
**      if a fatal error occurs, length will be set to -1
*/
char *zip_buffer(aClient *cptr, char *buffer, int *length, int flush)
{
  z_stream *zout = cptr->zip->out;
  int   r;

  if (buffer)
    {
      /* concatenate buffer in cptr->zip->outbuf */
      memcpy((void *)(cptr->zip->outbuf + cptr->zip->outcount), (void *)buffer,
             *length );
      cptr->zip->outcount += *length;
    }
  *length = 0;

#if 0
  if (!flush && ((cptr->zip->outcount < ZIP_MINIMUM) ||
                 ((cptr->zip->outcount < (ZIP_MAXIMUM - BUFSIZE)) &&
                  CBurst(cptr))))
	/* Implement this? more efficient? or not? -- Syzop */
#else
  if (!flush && (cptr->zip->outcount < ZIP_MINIMUM))
#endif
    return((char *)NULL);

  zout->next_in = (Bytef *) cptr->zip->outbuf;
  zout->avail_in = cptr->zip->outcount;
  zout->next_out = (Bytef *) zipbuf;
  zout->avail_out = ZIP_BUFFER_SIZE;

  switch (r = deflate(zout, Z_PARTIAL_FLUSH))
    {
    case Z_OK:
      if (zout->avail_in)
        {
          /* can this occur?? I hope not... */
          sendto_realops("deflate() didn't process all available data!");
        }
      cptr->zip->outcount = 0;
      *length = ZIP_BUFFER_SIZE - zout->avail_out;
      return zipbuf;

    default: /* error ! */
      sendto_realops("deflate() error(%d): %s", r, (zout->msg) ? zout->msg : "?");
      *length = -1;
      break;
    }
  return((char *)NULL);
}

#endif  /* ZIP_LINKS */
